crates/runtime/src/model/eval.rs

```diff
@@ -0,0 +1,441 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use crate::datafusion::{SPICE_DEFAULT_CATALOG, SPICE_DEFAULT_SCHEMA};
+
+use super::{eval_scorer::Scorer, DataFusion};
+use arrow::array::{
+    Array, ArrayRef, ListArray, RecordBatch, StringArray, StringViewArray, StructArray,
+};
+use arrow_schema::ArrowError;
+use async_openai::{
+    error::OpenAIError,
+    types::{
+        ChatChoice, ChatCompletionRequestMessage, ChatCompletionRequestUserMessageArgs,
+        CreateChatCompletionRequest, CreateChatCompletionRequestArgs,
+    },
+};
+
+use datafusion::sql::TableReference;
+use futures::TryStreamExt;
+use llms::chat::Chat;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use snafu::{ResultExt, Snafu};
+use spicepod::component::eval::Eval;
+use std::{collections::HashMap, sync::Arc};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to query eval dataset '{dataset_name}': {source}"))]
```

Comment:
Is there any action a user can take for any of these errors?

Reply:
Yes, I should add.